### PR TITLE
Update inventoryApiService.lua

### PR DESCRIPTION
**Pull request by systemNEO**
Fire the event server internally, too.

# Pull request template
> [!IMPORTANT]
> Please complete all fields. PRs will not be merged if any fields are incomplete. Be respectful, and keep in mind that it may take some time for your PR to be reviewed.
>
> Bear in mind refactors are up to the developers and not it's contributors after all we are the ones giving support when needed. if they are big changes I suggest you to release it under your name instead.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### Motive for This Pull Request
To trigger the event within the server it is possible to use the event by other scripts directly within the server instead of listening the event by the clients and re-fire the result to the server (which causes traffic).

### Provide a brief explanation of why these changes are being proposed and what they aim to achieve.
I want a solution to listen to "custom inventory open" on server side directly.

### Explain the necessity of these changes and how they will impact the framework or its users.
As developer which uses the framework it will help to build a performant solution to keep track of opened custom inventories. Currently I only can listen via clients for that event and trigger a server event to inform the server in my scripts. This is an unnecessary detour that would only cause additional traffic.

### Please describe the tests you have conducted to verify your changes. Provide instructions so we can reproduce these tests. I have tested it locally to ensure the event was thrown server internally whithout errors.

- [x] Tested with latest vorp scripts
- [x] Tested with latest artifacts

## Notes if any
The background is as follows: We have a quest script. In the script, an item is placed in the inventory with which you can open the quest log. We want to prevent this item from being moved from the player's inventory to any other inventory. There is the command "blackListCustomAny", but you have to know the custom inventories already, because you have to execute this command for every custom inventory that exists (if you know it). If we are informed via an event when any custom inventory is opened, we can execute the command on demand for the respective custom inventory. This means that we do not need to know every custom inventory in advance, but we can blacklist the item in every custom inventory. It is a pity that there is no function that restricts an item to the player inventory in general, this would be the better way, I am aware of that. If there is a better solution I would be thankfull.

### CODE REVIEW

**Comment by outsider31000** [comment]
Hi there!
I was thinking of using statebags to let server or client know that inventory is open either custom or main. I think would be better solution.
But from what you explained that you need this for, would this not work for you?

```lua
-- if the item was removed then remove quest from player
AddEventHandler("vorp_inventory:Server:OnItemRemoved",function(data)
 local source = data.source
  print(json.encode(data),{ident = true}) -- contains info about the item like name etc
end)
```
I think this would work for you you could cancel the quest if item was removed?
Let me know if it does work best for you.

**Comment by systemNEO** [comment]
Hello Outsider,

thanks for the quick reply. The idea is good in principle. Unfortunately, however, it does not exactly solve my problem.

Unfortunately, the item is not a simple quest item that is needed to complete a quest. Instead, it is the quest log itself. It opens the quest log. The quest log item is always placed in the player's inventory if the player has at least one active quest. If the player moves the quest log item to another inventory and relogs or starts the next quest step, our script checks whether the quest log item is in the player's inventory. If not, a quest log item is placed in it again. 

This is not the best solution because we cannot see where the player has moved the quest log item to and therefore have to create new ones in the player inventory. This will eventually lead to the player having quest log items in one of his custom inventories, but also in his player inventory. We want to prevent this. To prevent this, the item would have to be configurable so that it cannot be removed or moved from the player inventory. 

That's why I had the idea of simply blacklisting the quest log item in all custom inventories. But that only works if you know all the custom inventories, which you don't, because other scripts also add new ones on demand. However, the "vorp_inventory:OpenCustomInv" event allows me to find out what the player is currently opening and quickly blacklist the item there, just in time.

Basically, it would be good to be able to configure an item so that a) it cannot be removed and b) it cannot be moved to other inventories.

An alternative would be to design the appropriate event when moving or throwing away an item so that the action can be canceled via a return value.

As long as that's not possible, I would build the workaround by blacklisting the quest log item in it when opening a custom inventory. Therefore, it would be great if "vorp_inventory:OpenCustomInv" were also fired on the server side. Because then I don't have to know all the custom inventories and I don't have to go the long way round via the client and can save traffic.

Sorry for the long text.

**Comment by systemNEO** [comment]
PS: Is `vorp_inventory:Server:OnItemCreated` always fired when an item is placed in the player inventory? 

If so, that would also be an option: Then I would check every time a quest log item is added to the player inventory to see if there is more than one quest log item in the player inventory. If so, I would simply delete the surplus.

**Comment by outsider31000** [comment]
you can see in the docs for events https://vorpcore.github.io/VORP_Documentation/api/inventory#listeners

this event `OnItemRemoved` is fired when item is droppped moved or given , basically when is removed from main inv
then you have OnItemCreated as well everytime an item is created on main inventory  it fires.

> Basically, it would be good to be able to configure an item so that a) it cannot be removed and b) it cannot be moved to other inventories.

if you do in items table can_remove to 0 will it not work ?

**Comment by systemNEO** [comment]
Hey,

> this event OnItemRemoved is fired when item is droppped moved or given , basically when is removed from main inv
then you have OnItemCreated as well everytime an item is created on main inventory it fires.

I will test that.

> if you do in items table can_remove to 0 will it not work ?
Try that if its 0 it should not be allowed to remove item.

This seems to prevent dropping an item or other players can take that item from other players. But moving items into another inventories still work.

> I think what you want is something that hooks and sets an action before the items is moved from main Inventory right? so that you can cancel it etc

That idee would be very nice. Not just for this purpose, but in principle.

Thanks for help. I close the pull request for now. If I can not work with the ideas found here I simply will reopen the pull request.

**Comment by outsider31000** [comment]
yeah no problem feel free to open again if those dont work for you.
